[Aio] Implement server interceptor for unary unary call

ZHmao opened the pull request:
Support for interceptors for the unary-unary call at the server-side.

Fix #21914
Ref #20482

ZHmao commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi:
```diff
@@ -15,6 +15,7 @@
 
 import inspect
 import traceback
+import functools
```
I inject some parameters at here: https://github.com/grpc/grpc/pull/22032/files/6fef56573e9a0347c33c5aea4bc57ab625b0c6ea#diff-b53c077f7911f53dc8ce7656e16d35cdR222

ZHmao commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi:
```diff
@@ -214,15 +215,34 @@ cdef class _ServicerContext:
 
 
-cdef _find_method_handler(str method, tuple metadata, list generic_handlers):
+async def _run_interceptor(object interceptors, object query_handler,
```
I have no idea about the type name of an iterator.

ZHmao replied:
I'll update it, use `Iterator` .

ZHmao replied:
Could we use `Iterator` in `typing` ?

pfreixes replied:
Not sure if iterator is a type provided by Cython, like `tuple`. If not, `object` as a generic type would be the way to go.

ZHmao replied:
I found an `iterator` at here[1], dose it mean we can use it?

[1] https://github.com/cython/cython/tree/master/Cython/Includes/cpython

pfreixes replied:
could be! give it a try :)

ZHmao replied:
When I changed the type to iterator, the compiling failed, it shows:
> \'iterator\' is not a type identifier
There is no iterator type in cython.

ZHmao commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi:
```diff
@@ -214,15 +215,34 @@ cdef class _ServicerContext:
+
+async def _find_method_handler(str method, tuple metadata, list generic_handlers,
+                          tuple interceptors):
+    def query_handlers(handler_call_details):
```
We will need to iterate them in _run_interceptors[1] if we use generator at here.
Current version will make _run_interceptors simpler.

[1] https://github.com/grpc/grpc/pull/22032/files/6fef56573e9a0347c33c5aea4bc57ab625b0c6ea#diff-b53c077f7911f53dc8ce7656e16d35cdR218

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/interceptor_test.py:
```diff
@@ -685,6 +685,110 @@ class Interceptor(aio.UnaryUnaryClientInterceptor):
+
+        with self.assertRaises(aio.AioRpcError):
+            server_target, _ = await start_test_server(
+                interceptors=(InvalidInterceptor(),))
```
a `Channel` with an invalid interceptor is not allowed to be instantiated [1 ], having the feeling that we would need to do the same with the server.

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/experimental/aio/_channel.py#L317

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/interceptor_test.py:
```diff
@@ -685,6 +685,110 @@ class Interceptor(aio.UnaryUnaryClientInterceptor):
+            # in the right order.
+            self.assertSequenceEqual(['log1:intercept_service',
+                                      'log2:intercept_service',], self._record)
+            self.assertIsInstance(response, messages_pb2.SimpleResponse)
```
Maybe we can have another test - `test_response_ok` ? - for checking that the response is looking as we would be expecting by checking the `status` and the response content. With this specific test, we won't need to check at each test the response

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/interceptor_test.py:
```diff
@@ -685,6 +685,110 @@ class Interceptor(aio.UnaryUnaryClientInterceptor):
+            _LoggingServerInterceptor('log1', self._record),
+            conditional_interceptor,
+            _LoggingServerInterceptor('log2', self._record),
+        )
```
Having all of the interceptors added into the setup which they would be used later for all of the tests makes the readability a bit harder for me, TBH I would prefer to have each test case with its own server with its own interceptors for testing the different functionalities.

The same as you already did for the `test_invalid_interceptor` test case.

Having the feeling that having this kind of sharing, the same happen with the `_record` class attribute, we have a lot of chances of creating down-side effects between threads.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/interceptor_test.py:
```diff
@@ -685,6 +685,110 @@ class Interceptor(aio.UnaryUnaryClientInterceptor):
+    return _GenericServerInterceptor(intercept_service)
+
+
+class TestServerInterceptor(AioTestBase):
```
I would like to have a test for checking observability of the code responses sent by the handler, in the same way as we have for the channel [1]

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests_aio/unit/interceptor_test.py#L93

ZHmao replied:
Server interceptors are wrapped at `find_method_handler` , the interceptors won't get the status code of response.

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi:
```diff
@@ -214,15 +215,34 @@ cdef class _ServicerContext:
+        return await _run_interceptor(iter(interceptors), query_handlers,
+                                      handler_call_details)
+    else:
+        return query_handlers(handler_call_details)
```
Both `_run_interceptor` and `_find_method_handler` functions are on the data path, is it possible to simplify them? Or run the benchmark to check the impact on performance?

pfreixes replied:
what about doing something like:

```python
async def  _find_method_handler(...):

    for generic_handler in generic_handlers:
        method_handler = generic_handler.service(handler_call_details)
        if method_handler is not None:
            break

    if interceptors:
        return await _run_interceptor(....., method_handler)
    else:
        return method_handler
```

With this change theoretically the execution path when there is no interceptors should not be hammered.

WDYT? /cc @lidizheng

lidizheng replied:
It depends on whether we allow users to change the method.

If allowed, the method handler needs to be ran as last function in the chain, otherwise users' interception might not work.

If not allowed, this is a valid optimization.

lidizheng commented on src/python/grpcio_tests/tests_aio/unit/interceptor_test.py:
```diff
@@ -685,6 +685,110 @@ class Interceptor(aio.UnaryUnaryClientInterceptor):
+        return await self._fn(continuation, handler_call_details)
+
+
+def _filter_server_interceptor(condition, interceptor):
```
To improve the maintainability of test cases, we might want to add type annotations, especially for callbacks.

lidizheng commented on src/python/grpcio_tests/tests_aio/unit/interceptor_test.py:
```diff
@@ -685,6 +685,110 @@ class Interceptor(aio.UnaryUnaryClientInterceptor):
+                                      'log3:intercept_service',
+                                      'log2:intercept_service',],
+                                     self._record)
+
```
Should we add one more test case that the interceptor returns a different method handler?

ZHmao replied:
I couldn't understand this one, can you give me more details?

mehrdada commented:
One of my regrets with the original design for sync interceptors is allowing server-interceptors to change the path which massively limits the opportunity of optimization for handing over to handlers and is non-idiomatic-gRPC. Please make sure you at least revisit whether this flexibility is worth the cost.

lidizheng commented:
@mehrdada By

> ... which massively limits the opportunity of optimization for handing over to handlers...

Are you referring to the potential optimization in finding methods? Otherwise, it could be done in Cython or C++? Can you explain more about this topic?

pfreixes commented:
@ZHmao also consider that current PR has conflicts that are preventing us to run the whole CI, I would appreciate if you could resolve the conflicts.

Thanks!

mehrdada commented:
@lidizheng C-Core server supports registering RPC paths directly upfront as opposed to everything be "the generic handler".  Python API is clearly designed to be able to leverage it (see `ServiceRpcHandler` and `GenericRpcHandler`) but never did. If we know which handler is going to be invoked upfront, we can create the interceptor handling pipeline per path upfront and have C core or Cython via a proper efficient hashtable that can run outside GIL or whatever handle it efficiently before we hand it over to Python, basically letting slow application code to resolve which method handler is going to be invoked. With the generic server interceptor API, we give user flexibility to change it which is often overkill and we won't know the handler to dispatch to upfront.

ZHmao commented:
> @ZHmao also consider that current PR has conflicts that are preventing us to run the whole CI, I would appreciate if you could resolve the conflicts.
> 
> Thanks!

Sure, I will.

pfreixes commented:
@mehrdada @lidizheng having the chance of adding interceptors for specific handlers is something that could be really useful, not only because of performance but because the user has the chance of adding a piece of logic that is related to a specific path.

On the other hand, generic handlers are gonna be needed still for doing generic stuff like observability, tracing, etc ...

IMO the interceptors discussion with regards the design is not yet closed, we have two outstanding g questions that would need to be answered before making this feature as no longer experimental:

- The granularity of the interceptors. how does the current design work with the other arities? A design based on hooks would make everything easier? and more important less prone error? 
- Generic handlers vs route/path handlers. How we would implement both use cases? do the route/path handlers make sense on the client-side?

Right now we are on the way of doing an alpha release and this feature is marked as experimental, so I guess that we have still full freedom on breaking the API or extending it without having the friction of being in a beta/stable release.

mehrdada commented:
@pfreixes I am fine with releasing alpha or whatever as long as it can be changed. To clarify, my concern is a bit orthogonal to interceptors being able to attach to specific paths (although I am not necessarily a fan, it is a valid feature to debate).  The limitation that I like to see enforced is interceptor not being able to *alter* the path it is invoked on, i.e. return a different path than the one it's invoked on. It is valid for it to see the path. It might require more work and a more complex API, but it is alright if it can be invoked on certain paths and not others if they are known upfront, because you can still construct the handler pipelines upfront at server start, even if they are different per request path.

pfreixes commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi:
```diff
@@ -214,15 +215,34 @@ cdef class _ServicerContext:
-    return None
+    # interceptor
+    if interceptors:
+        return await _run_interceptor(iter(interceptors), query_handlers,
```
Should we return and not execute, the `_run_interceptor` coroutine? This would mean that the execution will eventually happen at the very last step and the interceptor for example in the case of a unary call will be here [1]

Indeed we could name this method like `_wrap_call_interceptor` or something like that.

Without doing this, future interceptors that would take care of observability for measuring the time that handler took won't work since they won't be able to measure the real execution of the handler.

Also moving this to the right place, will eventually will help us for extending the interceptors for having visibility of the servicer context [2] at the interceptor level.

The only friction point of doing this that I can see, if Im not missing something is the access to the metadata. So the interceptor won't be able to have direct access to the metadata and mutate them, but could have access to the service context that will make the metadata available for the interceptor.

WDYT @lidizheng should we aim for moving the execution of the interceptor to the very last step?

Also, I'm wondering if since we are not most likely commit to having this feature for the alpha release we could start thinking on how a nex iteration of the servers interceptors could be done for knowing the response code of the method handler. Would this change help us to do so? 

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi#L242

pfreixes commented on src/python/grpcio/grpc/experimental/aio/_server.py:
```diff
@@ -41,6 +42,13 @@ def __init__(self, thread_pool: Optional[Executor],
+                                                      ServerInterceptor)]
+            if invalid_interceptors:
+                raise ValueError('Interceptor must be ServerInterceptor, the '
+                                 f'following are invalid: {invalid_interceptors}')
```
I think that we committed to not use f-strings, even though the minimal target python version targeted for now is 3.6, which supports f strings.

@lidizheng f-strings yes or not?

lidizheng replied:
In Python, yes; in Cython, no.

If possible, I would like to use this Python 3 feature (simple is better than complex).

Cython doesn't understand f-string if it is compiled with Python < 3.6. We have learnt this in through test failures.

pfreixes commented:
@lidizheng and @ZHmao Im happy on having an iterative process for implementing the server interceptors, so having first implemented a version that works as the legacy stack is IMO already giving value.

So I would not have any inconvenience in approving this PR and later on have other PR for making the server interceptors more powerful. Indeed I would say that it will be the best thing to do.

But there is something that I do not understand, looking at this interceptor [1] that is the one used by one of our first adopters of gRPC that are using the legacy stack, seems that the server version allows you to at least:

- Catch exceptions returned by the handler
- Get the response returned by the handler
- Measure the time spent within the handler 

Looking at the PR seems that neither of the previous capabilities are ready to be used. am I wrong? what Im missing?

[1] https://github.com/opentracing-contrib/python-grpc/blob/325edbb83f08110e7d2ea1d1c453aaf61666b6e7/grpc_opentracing/_server.py#L139

ZHmao commented:
Hi @pfreixes , they used a new class `_InterceptorServer` to wrap the original server [1], and changed the behavior of `add_generic_rpc_handlers`, they added interceptor for every rpc handler.

[1] https://github.com/opentracing-contrib/python-grpc/blob/325edbb83f08110e7d2ea1d1c453aaf61666b6e7/grpc_opentracing/grpcext/_interceptor.py#L342

pfreixes commented:
@ZHmao wow, I missed that! So they did what we would like to have.... great example on how the users follow their own paths for circumventing the current limitations.

So in that case, if we are already meeting the same requirements that are met by the legacy stack take me 👍 but I would like to see a plan for designing a server interceptor that supersedes the current one and is paired to what _lightstep_ [1] has done.

PTAL to the sanity checks, most. of them are failing.

[1] https://github.com/opentracing-contrib/python-grpc/blob/325edbb83f08110e7d2ea1d1c453aaf61666b6e7/grpc_opentracing/grpcext/_interceptor.py#L342

lidizheng commented on src/python/grpcio_tests/tests_aio/unit/server_interceptor_test.py:
```diff
@@ -0,0 +1,144 @@
+# Copyright 2019 The gRPC Authors.
```
```suggestion
# Copyright 2020 The gRPC Authors
```

lidizheng commented on src/python/grpcio/grpc/experimental/aio/_server.py:
```diff
@@ -152,7 +160,8 @@ def __del__(self):
         """
-        self._loop.create_task(self._server.shutdown(None))
+        if hasattr(self, '_server'):
+            self._loop.create_task(self._server.shutdown(None))
```
👍

ZHmao commented:
Thanks for all your reviews.

lidizheng commented on src/python/grpcio_tests/tests_aio/unit/server_interceptor_test.py:
```diff
@@ -0,0 +1,152 @@
+        return await self._fn(continuation, handler_call_details)
+
+
+def _filter_server_interceptor(condition: Callable,
```
I missed this one in my last round of review. By adding type annotation, I mean what is the argument it accepts and returns. You can do this like:

```python
def atoi(a: str) -> int:
    return int(a)

atoi: Callable[[str], int]
```

Also, if possible, please annotate all other arguments in this file, not only this function.

ZHmao replied:
OK

ZHmao commented:
I have added type annotation for `server_interceptor_test.py`, and changed the return type of `continuation` from `grpc.RpcMethodHandler` to `Awaitable[grpc.RpcMethodHandler]` .

But it failed when I run `python -m pytype` .
The error message:
```
in _filter_server_interceptor: Function _GenericInterceptor.__init__ was called with the wrong arguments [wrong-arg-types]
  Expected: (self, fn: Callable[[Callable[[grpc.HandlerCallDetails], Awaitable[grpc.RpcMethodHandler]], grpc.HandlerCallDetails], Awaitable[grpc.RpcMethodHandler]])
  Actually passed: (self, fn: Callable[[Callable[[grpc.HandlerCallDetails], Awaitable[grpc.RpcMethodHandler]], grpc.HandlerCallDetails], grpc.RpcMethodHandler])
```

cc @lidizheng

lidizheng commented:
@ZHmao I'm not sure I understand the question here. Did you mean the `pytype` has a deduction error on the types? In that case, you can post it as an issue to `pytype`. Otherwise, the error message seems clear that there might be a confusion in the implementation.

(As I said, complex callbacks are usually confusing and needs proper typing.)

ZHmao commented:
I have written an example code and it shows the same error:
`
 in go: Function bar was called with the wrong arguments [wrong-arg-types]
         Expected: (fun: Callable[[str], Awaitable[str]])
  Actually passed: (fun: Callable[[str], str])
`


```
from typing import Callable, Awaitable


async def foo(a: str) -> str:
    return a


async def bar(fun: Callable[[str], Awaitable[str]]) -> str:
    return await fun('a')


async def go() -> None:
    await bar(foo)

```

I don't know what's wrong, do you have any advice?

lidizheng commented:
@ZHmao Thanks for the convincing code snippet, good find! It looks like `pytype` has a bug, that it unable to express async functions correctly. In the generated `.pyi` file, it recognized that `foo` is an async function that returns a coroutine. However, when you passes `foo` as an argument, `pytype` complains.

```python
# Generated .pyi file
def bar(fun: Callable[[str], Awaitable[str]]) -> Coroutine[Any, Any, str]: ...
def foo(a: str) -> Coroutine[Any, Any, str]: ...
def go() -> Coroutine[Any, Any, None]: ...
```

If you have some free cycle and want to be a contributor of `pytype`, feel free to open an issue or post a PR directly to https://github.com/google/pytype. You can find their unit test here: https://github.com/google/pytype/blob/master/pytype/tests/py3/test_coroutine.py.

On the other hand, to move this PR forward, there are two (hacky) ways to solve the `pytype` complain.
1. Use `Union` for the return value: `Union[grpc.RpcMethodHandler, Awaitable[grpc.RpcMethodHandler]]`;
2. Use `Any` for the return value.

@pfreixes Have you encountered this bug before?

ZHmao commented:
I'll use `Union` to move this PR forward, and open an issue or post a PR on pytype later.

ZHmao commented:
FYI, when I tried to fix this issue with `Union`, it failed again. Then I wrote an example more like ours.😅
```
from typing import Callable, Awaitable, Union


async def foo(a: str) -> str:
    return a


class Bar:

    def __init__(self, fun: Callable[[str], Union[Awaitable[str], str]]) -> None:
        self.fun = fun

    async def bar(self) -> str:
        return await self.fun('a')


async def go() -> None:
    Bar(foo)

```

Error message:
```
in bar: bad option in return type [bad-return-type]
           Expected: Awaitable
  Actually returned: str

```

But type `Any` is OK in this situation. So I'll use `Any` to move this PR forward.

cc @lidizheng @pfreixes

lidizheng commented:
@ZHmao Thanks for the update. This definitely looks like an issue for `pytype`.

ZHmao commented:
Can we merge it? Or I need to make all checks pass before we can merge it?

lidizheng commented:
Known failures: https://github.com/grpc/grpc/issues/22208 https://github.com/grpc/grpc/issues/19834
